### PR TITLE
Transformations for hsigmoid op

### PR DESCRIPTION
**Pull request by iimironov**

### CODE REVIEW

**Review by ilyachur — COMMENTED** [review]
Could you add a tests for this operation?

@iimironov Please the measure impact from evaluate method on nGraph binary size.

File: ngraph/core/include/ngraph/opsets/opset5_tbl.hpp [hunk 1]
```diff
@@ -166,4 +166,5 @@ NGRAPH_OP(Swish, ngraph::op::v4)
 // New operations added in opset5
 NGRAPH_OP(LSTMSequence, ngraph::op::v5)
 NGRAPH_OP(GRUSequence, ngraph::op::v5)
-NGRAPH_OP(RNNSequence, ngraph::op::v5)
\ No newline at end of file
+NGRAPH_OP(RNNSequence, ngraph::op::v5)
+NGRAPH_OP(HSigmoid, ngraph::op::v5)
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/opsets/opset5_tbl.hpp:170]: Please, insert the op in the sorted name order to the section with opset5 ops.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]
Please, fix the code style.

**Comment by lazarevevgeny** [comment]
We also need fusing transformations and tests for them.

**Comment by lazarevevgeny** [comment]
@iimironov , have you measured the impact on binary size requested by @ilyachur ?

**Comment by iimironov** [comment]
> Could you add a tests for this operation?
> 
> @iimironov Please the measure impact from evaluate method on nGraph binary size.


With hsigmid
-rwxr-xr-x 1 imironov inn_avc_ocl_qa  19078424 Oct  9 16:24 libngraph_backend.so
-rw-r--r-- 1 imironov inn_avc_ocl_qa  15492478 Oct  9 16:21 libngraph_builders.a
-rw-r--r-- 1 imironov inn_avc_ocl_qa 109895866 Oct  9 16:28 libngraphFunctions.a
-rw-r--r-- 1 imironov inn_avc_ocl_qa   2947466 Oct  7 12:26 libngraph_reference.a
-rwxr-xr-x 1 imironov inn_avc_ocl_qa  98444624 Oct  9 16:23 libngraph.so
-rw-r--r-- 1 imironov inn_avc_ocl_qa  10800876 Oct  9 16:25 libngraph_test_util.a

Without hsigmoid
-rwxr-xr-x 1 imironov inn_avc_ocl_qa  19078424 Oct  9 16:57 libngraph_backend.so
-rw-r--r-- 1 imironov inn_avc_ocl_qa  15492478 Oct  9 16:54 libngraph_builders.a
-rw-r--r-- 1 imironov inn_avc_ocl_qa 109893690 Oct  9 16:59 libngraphFunctions.a
-rw-r--r-- 1 imironov inn_avc_ocl_qa   2947466 Oct  7 12:26 libngraph_reference.a
-rwxr-xr-x 1 imironov inn_avc_ocl_qa  98212584 Oct  9 16:56 libngraph.so
-rw-r--r-- 1 imironov inn_avc_ocl_qa  10800876 Oct  9 16:58 libngraph_test_util.a

**Comment by lazarevevgeny** [comment]
> > Could you add a tests for this operation?
> > @iimironov Please the measure impact from evaluate method on nGraph binary size.
> 
> With hsigmid
> -rwxr-xr-x 1 imironov inn_avc_ocl_qa 19078424 Oct 9 16:24 libngraph_backend.so
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 15492478 Oct 9 16:21 libngraph_builders.a
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 109895866 Oct 9 16:28 libngraphFunctions.a
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 2947466 Oct 7 12:26 libngraph_reference.a
> -rwxr-xr-x 1 imironov inn_avc_ocl_qa 98444624 Oct 9 16:23 libngraph.so
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 10800876 Oct 9 16:25 libngraph_test_util.a
> 
> Without hsigmoid
> -rwxr-xr-x 1 imironov inn_avc_ocl_qa 19078424 Oct 9 16:57 libngraph_backend.so
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 15492478 Oct 9 16:54 libngraph_builders.a
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 109893690 Oct 9 16:59 libngraphFunctions.a
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 2947466 Oct 7 12:26 libngraph_reference.a
> -rwxr-xr-x 1 imironov inn_avc_ocl_qa 98212584 Oct 9 16:56 libngraph.so
> -rw-r--r-- 1 imironov inn_avc_ocl_qa 10800876 Oct 9 16:58 libngraph_test_util.a

I have a doubt that a simple reference implementation function resulted in 200K+ file size change. Did you remove just the evaluate method and everything related to it?

**Comment by iimironov** [comment]
> evaluate

with  evaluate
98444624 libngraph.so

without  evaluate
98423864 libngraph.so

diff 20 760

**Comment by lazarevevgeny** [comment]
@iimironov, the OpenVINO-ONNX fails because of the following error: "src/ngraph/opset5/ops.py:68:77: W292 no newline at end of file"

Please, fix.

**Comment by iimironov** [comment]
> @iimironov, the OpenVINO-ONNX fails because of the following error: "src/ngraph/opset5/ops.py:68:77: W292 no newline at end of file"
> 
> Please, fix.

Fixed

**Comment by lazarevevgeny** [comment]
@iimironov , please, merge changes from master branch. Otherwise you will not be able to pass the pre-commit.

File: inference-engine/src/transformations/src/transformations/hsigmoid_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/hsigmoid_fusion.cpp:1]: We need also a decomposition transformation for HSigmoid for the plugins which do not support HSigmoid directly. For now it must be enabled for all plugins.

File: ngraph/core/include/ngraph/opsets/opset5_tbl.hpp [hunk 3]
```diff
@@ -163,11 +163,11 @@ NGRAPH_OP(SoftPlus, ngraph::op::v4)
 NGRAPH_OP(Swish, ngraph::op::v4)
 
 // New operations added in opset5
+NGRAPH_OP(NonMaxSuppression, ngraph::op::v5)
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/opsets/opset5_tbl.hpp:166]: Please, revert this change.

**Review by ilyachur — COMMENTED** [review]
Looks like the description of this PR isn't consistent with content. Am I not right?

**Comment by iimironov** [comment]
> Looks like the description of this PR isn't consistent with content. Am I not right?

We separated this PR on two. One content only ngraph op and other content transformations and python api.
This PR contain  transformations and python api.

File: inference-engine/src/transformations/include/transformations/common_optimizations/hsigmoid_fusion.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include <utility>
+
+#include <transformations_visibility.hpp>
+#include <ngraph/pass/graph_rewrite.hpp>
+
+namespace ngraph {
+namespace pass {
+
+class TRANSFORMATIONS_API HSigmoidFusion;
+class TRANSFORMATIONS_API HSigmoidFusionWithReluDiv;
```
Comment (ilyachur) [inline inference-engine/src/transformations/include/transformations/common_optimizations/hsigmoid_fusion.hpp:17]: Why do we need to export all matchers if we use only `HSigmoidFusion` class in other libraries?

I propose to export only `HSigmoidFusion`.

File: inference-engine/src/transformations/include/transformations/common_optimizations/hsigmoid_fusion.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        add_matcher<ngraph::pass::HSigmoidFusionWithoutRelu>();
+        add_matcher<ngraph::pass::HSigmoidFusionWithClamp>();
+    }
+};
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief HSigmoidFusion transformation replaces a sub-graph (x * (min(Relu(x + 3), 6))) / 6 with a HSigmoid op.
+ */
+class ngraph::pass::HSigmoidFusionWithReluDiv: public ngraph::pass::MatcherPass {
```
Comment (ilyachur) [inline inference-engine/src/transformations/include/transformations/common_optimizations/hsigmoid_fusion.hpp:45]: Also if you move constructor of  `HSigmoidFusion` class to cpp we can hide (move to cpp) all matcher passes.

File: inference-engine/src/transformations/src/transformations/common_optimizations/hsigmoid_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            ngraph::replace_node(m.get_match_root(), hsigmoid);
+            return true;
+        };
+
+        auto m = std::make_shared<ngraph::pattern::Matcher>(mul_first, "HSigmoidWithClampFusion");
+        register_matcher(m, callback);
+    }
+};
+
+ngraph::pass::HSigmoidFusion::HSigmoidFusion() {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/hsigmoid_fusion.cpp:236]: Please move it to hpp file as we do in other transformations

File: inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+ngraph::pass::HSigmoidDecomposition::HSigmoidDecomposition() {
+    // Decomposes HSigmoid(x) op into sub-graph (min(Relu(x + 3), 6) * const(1/6)
+    auto hsigmoid = ngraph::pattern::wrap_type<opset5::HSigmoid>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hsigmoid_node = pattern_to_output.at(hsigmoid).get_node_shared_ptr();
+
+        if (m_transformation_callback(hsigmoid_node)) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp:23]: Do we have plugins that disable this pass by condition?

File: inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::HSigmoidDecomposition, "HSigmoidDecomposition", 0);
+
+ngraph::pass::HSigmoidDecomposition::HSigmoidDecomposition() {
+    // Decomposes HSigmoid(x) op into sub-graph (min(Relu(x + 3), 6) * const(1/6)
+    auto hsigmoid = ngraph::pattern::wrap_type<opset5::HSigmoid>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hsigmoid_node = pattern_to_output.at(hsigmoid).get_node_shared_ptr();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp:21]: Just use `m.get_match_root()` here

File: inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+        auto input_type = hsigmoid_node->input_value(0).get_element_type();
+        auto add_constant = ngraph::opset5::Constant::create(input_type, ngraph::Shape{}, {3.0});
+        auto add = std::make_shared<ngraph::opset5::Add>(hsigmoid_node->input_value(0), add_constant);
+        auto relu = std::make_shared<ngraph::opset5::Relu>(add);
+        auto min_constant = ngraph::opset5::Constant::create(input_type, ngraph::Shape{}, {6.0});
+        auto min = register_new_node<ngraph::opset5::Minimum>(relu, min_constant);
+        auto mul = std::make_shared<ngraph::opset5::Multiply>(hsigmoid_node->input_value(0), min);
+
+        mul->set_friendly_name(m.get_match_root()->get_friendly_name());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp:35]:
```suggestion
        mul->set_friendly_name(hsigmoid->get_friendly_name());
```

File: inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        ngraph::copy_runtime_info(hsigmoid_node,
+                                  {add_constant, add, relu, min_constant, min, mul});
+        ngraph::replace_node(m.get_match_root(), mul);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/op_conversions/hsigmoid_decomposition.cpp:38]:
```suggestion
        ngraph::replace_node(hsigmoid, mul);
```

File: inference-engine/tests/functional/inference_engine/transformations/hsigmoid_decomposition_test.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+        ngraph::pass::Manager manager;
+        manager.register_pass<ngraph::pass::InitNodeInfo>();
+        manager.register_pass<ngraph::pass::HSigmoidDecomposition>();
+        manager.run_passes(f);
+        ASSERT_NO_THROW(check_rt_info(f));
+    }
+
+    {
+        auto input = std::make_shared<ngraph::opset5::Parameter>(ngraph::element::f16, ngraph::PartialShape::dynamic(1));
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/hsigmoid_decomposition_test.cpp:37]: Why you have f16 type for reference net?

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Review by ilyachur — APPROVED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**